### PR TITLE
XDOCKER-69: Expose tomcat server.xml to ensure reverse proxy compatibility

### PR DESCRIPTION
**Pull request by zonradkuse**
As described in https://www.xwiki.org/xwiki/bin/view/Documentation/AdminGuide/Installation/InstallationWAR/InstallationTomcat/ tomcat needs some configuration to make SSL work properly.

Hence, I took the tomcat docker image's server.xml and modified it accordingly. This change should not affect users not using reverse proxies.

### CODE REVIEW

**Comment by zonradkuse** [comment]
It has been nearly a month and the XWiki development continues. I would appreciate comments on my PR or even a merge into master.

**Comment by vmassol** [comment]
Hi @zonradkuse Yeah, thanks for your PR and sorry for not having answered. July and August are very slow months and everyone is on holidays. I had marked your PR in my mailbox but haven't reached it yet. I'll have a look ASAP.

**Comment by zonradkuse** [comment]
That's fine. I was just afraid that it had gone missing in a bunch of emails. Thank you!

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,5 @@
-#Wed Apr 26 18:18:58 CEST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.4.1-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.4.1-bin.zip
```
Comment (vmassol) [inline gradle/wrapper/gradle-wrapper.properties]: Hi. Is that a mistake? Why make this change for this PR? :)

**Comment by vmassol** [comment]
@zonradkuse why the changes brought to gradle? I don't think that should be included in this PR, right?

Also, I see you're adding a server.xml config. I have 2 questions regarding this:
1) Does it change something for existing users of the XWiki Docker image who will upgrade? Will they still have the same config as they had before for HTTP for example? Does it simply add HTTPS but keeps HTTP?
2) I'm worried that this is going to be maintenance intensive since it means that we need to diff the default server.xml provided in the tomcat base image all the time since this base image could make modification to the default server.xml at any time. Any idea how to handle this? Or maybe the default server.xml is never changed and thus there's little maintenance? :)

Thanks for your PR! Once I understand better the above, I should be able to merge it quickly.

**Comment by vmassol** [comment]
BTW we need a JIRA issue on jira.xwiki.org + the reference to the issue in the commit comment. See https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HRule:AlwaysputaJIRAissuereferenceincommitmessages

I've found https://jira.xwiki.org/browse/XDOCKER-69 which could be the issue that you implemented, right?

**Comment by zonradkuse** [comment]
I didn't open an issue, however, I do only add SSL by using a reverse proxy in order to affect the least possible amount of users. XWiki's build in SSL is not enabled on purpose. This would need more complicated changes. I think SSL through a reverse proxy is the most elegant way to solve this issue. If you don't mind, I will use the issue reference anyway.

Yes, the gradle change is a mistake I didn't notice. I can't remember touching the gradle whatsoever, tho.

1. Users upgrading will have the same tomcat config with the addition of `<Valve className="org.apache.catalina.valves.RemoteIpValve">{...]</Valve>`. This change won't affect them if tomcat isn't used behind a reverse proxy. It does only add HTTPS if a reverse proxy is configured accordingly and keep HTTP at all times.

2. Apache Tomcat is a very stable project in my opinion. I wouldn't expect any changes if the base image isn't upgraded to a more recent version. And even then it is unlikely to change. The most recent commit on server.xml is dated 2016 only updating comments. https://github.com/apache/tomcat/commits/master/conf/server.xml

If you agree I will start amending the PR and removing the gradle change

**Comment by vmassol** [comment]
Thanks for your answers.

> This change won't affect them if tomcat isn't used behind a reverse proxy. 

Quite a number of XWiki users are using the docker image behind a reverse proxy so we would need to document this somewhere or to take it into account automatically. BTW one idea is to not change the provided server.xml but to write some shell script that modifies it on the first execution. We already have such a script and know when it's executed the first time, see https://github.com/xwiki-contrib/docker-xwiki/blob/master/template/xwiki/docker-entrypoint.sh

WDYT?

Thanks

**Comment by zonradkuse** [comment]
I like the idea of reusing the mechanism you already use for the xwiki configuration. However, I imagine it quite tricky to modify a xml file using bash. I'd be fine with simply exposing the server.xml similarly to all other configs. This doesn't enable SSL automatically but rather enables the user to make modifications in order to enable SSL.

If you agree I would rather close this PR, and open a second one which exposes server.xml of the underlying tomcat using the existing copy mechanism. The only thing to document is then that a user should check every few updates for tomcat server.xml updates.

**Comment by zonradkuse** [comment]
Hi @vmassol,

I applied a few changes to docker-endpoint.sh to expose server.xml. The mechanism should now be able to expose arbitrary files from inside the image.

**Comment by vmassol** [comment]
@zonradkuse ok I see, thanks for the changes. So you now expose the server.xml file so that users can configure it the way they want. Just to be sure I understand: your updated PR does less than the previous one, and it leaves the config of server.xml to the user. Correct?

Now I still have some doubts:
* What is the advantage of your PR vs using a docker run command which mounts the Tomcat `/usr/local/tomcat/conf/` directory? Isn't this just a matter of documenting it?
* IMO we could make it simpler for XWiki users by offering some arguments to the xwiki docker image for configuring SSL. Then it would be the job of https://github.com/xwiki-contrib/docker-xwiki/blob/master/template/xwiki/docker-entrypoint.sh to perform the config. TBH I've never configured xwiki for SSL so I don't know the setup but it seems to me that it wouldn't be that hard to do with some `sed` calls to modify Tomcat's `server.xml`. Now this second point could be a way to implement XDOCKER-69 (based on my last comment in that issue).

WDYT?

**Comment by zonradkuse** [comment]
Hi @vmassol 

> What is the advantage of your PR vs using a docker run command which mounts the Tomcat /usr/local/tomcat/conf/ directory

First of all you already commented on this in the jira issue. This is one of the reasons why I didn't do it first place. One advantage of this PR is that it only exposes server.xml for now where most changes happen if there are any needed for XWiki.

It is true that it would be easier for users, however, it is also considerably more effort since there are multiple ways of configuring XWiki for SSL. It is also not exactly the aim of this PR to implement full SSL support. My goal is having a way of configuring SSL in the end, however for that purpose I only need server.xml exposed. Now there is a way without exposing most of tomcat's configuration.

**Comment by zonradkuse** [comment]
>  Just to be sure I understand: your updated PR does less than the previous one, and it leaves the config of server.xml to the user. Correct?

Exactly. :-)

**Comment by vmassol** [comment]
>  however for that purpose I only need server.xml exposed. Now there is a way without exposing most of tomcat's configuration.

I get that. However, this is about exposing Tomcat's config in general. It just happens that for your specific use case you just need to expose server.xml, but for other use cases, any part of Tomcat's config could need to be changed. It's configuration after all.

It's like saying that you only need to make changes to xwiki's xwiki.cfg file but not xwiki.properties so we shouldn't expose xwiki.properties. But YMMV and depending on your use case you may need to make changes to more than one Tomcat config file.

So for me it makes total sense that if a user wants to change Tomcat's config, he should mount a volume to it. And the beauty is that it's independent of XWiki!

Now AFAIK Docker accepts mounting both directories and files (https://docs.docker.com/storage/bind-mounts/) so I don't see why a user wouldn't be able to mount just server.xml if that's the need.

Acceptable? :)

**Comment by zonradkuse** [comment]
I still think this should be somehow managed by XWiki since there could be changes to the image breaking paths. Your solutions simply loads the responsibility to the user if I understand you correctly.

> It's like saying that you only need to make changes to xwiki's xwiki.cfg file but not xwiki.properties so we shouldn't expose xwiki.properties
I don't think so. This is a XWiki docker image. There is no real need to expose anything but tomcat certain configuration files since it is a third-party project. XWiki should definitely expose all its very own configuration.

**Comment by vmassol** [comment]
@zonradkuse You haven't responded to the last part of my comment:

> Now AFAIK Docker accepts mounting both directories and files (https://docs.docker.com/storage/bind-mounts/) so I don't see why a user wouldn't be able to mount just server.xml if that's the need.

WDYT about that? Thx

**Comment by zonradkuse** [comment]
Similarly. If the underlying path in the container changes, it requires the user to change a path. Maybe I misunderstand your solution. You don't see any changes for that purpose within the XWiki docker image?!

**Comment by zonradkuse** [comment]
Also, mounting some server.xml into the container requires knowledge of the inner workings of the container. For example, the default port. By exposing the default server.xml the user doesn't need to lookup the internal port.

**Comment by zonradkuse** [comment]
I added a volume for server.xml to our project. If anyone is running into this I am going to leave you a link here. I still think that it is not an elegant solution to use volumes since it forces me to provide a server.xml instead of modifying one. Effectively, I am modifying the default tomcat one, though, ignoring any image specific settings.

https://github.com/studieren-ohne-grenzen/sog-ansible/blob/master/roles/xwiki/templates/docker-compose.yml.j2

**Comment by vmassol** [comment]
You’re closing the PR too fast ;) We’re just discussing and I’m not against your idea. Just need to see the pros and cons of each approach. Give me a few more days (I’m off since yesterday and for 10 days).

**Comment by zonradkuse** [comment]
I'm sorry, I got the impression that this is pretty much settled. I will reopen it for now. Have a good time off! :)

**Comment by vmassol** [comment]
Back from holidays. Trying to respond to all your points.

> I still think this should be somehow managed by XWiki since there could be changes to the image breaking paths. Your solutions simply loads the responsibility to the user if I understand you correctly.

So I agree with you that it's not great to have the image user know the location of the tomcat conf path and we could imagine (although it's not going to happen in practice) that we could change where the tomcat conf files are located.

We could fix this by creating a symbolic link inside the image to the tomcat conf directory and then users would mount this fixed location.

> Also, mounting some server.xml into the container requires knowledge of the inner workings of the container. For example, the default port. By exposing the default server.xml the user doesn't need to lookup the internal port.

I don't understand this part because the idea woule be to expose the content of the tomcat conf directory to let users make modifications to it when needed. It's not force users to provide a full fledged tomcat configuration. Technically this is https://docs.docker.com/engine/reference/builder/#volume

IMO this provides the best of both worlds:
* Ability to modify any part of Tomcat's configuration (whether you need to change the number of threads, or the port, or...)
* Fixed location so that even if the location of Tomcat changes, it'll still work. This second point is really minor IMO since the likeliness that it changes is so low and even if it did, we could update the documentation.

WDYT?

**Comment by zonradkuse** [comment]
Hi Vincent,

step by step... 
> We could fix this by creating a symbolic link inside the image to the tomcat conf directory and then users would mount this fixed location.

This does only expose the symlink, doesn't it? All changes would (if they happen) inside a container modifying a non-exposed file which will be restored after the container is restarted. I am not sure whether editing would actually work.

> I don't understand this part because the idea woule be to expose the content of the tomcat conf directory to let users make modifications to it when needed. It's not force users to provide a full fledged tomcat configuration. Technically this is https://docs.docker.com/engine/reference/builder/#volume

Yes, if you expose a directory this should be working that way. I am not sure what my problem was. Sorry.

**Comment by vmassol** [comment]
> This does only expose the symlink, doesn't it? All changes would (if they happen) inside a container modifying a non-exposed file which will be restored after the container is restarted. I am not sure whether editing would actually work.

What I meant is that the user would write something like `docker run -v /..path/tomcatconf:/usr/local/tomcatconf`. Inside the image the `/usr/local/tomcatconf` directory would be a symlink to `/usr/local/tomcat/conf`.

Now, I really think this is overkill and we don't need this. We can just have users write: `docker run -v /..path/tomcat/conf:/usr/local/tomcat/conf` IMO. We already expose several paths if you check the doc at https://github.com/xwiki-contrib/docker-xwiki/blob/master/README.md, like `/var/lib/mysql`, `/var/lib/postgresql/data`, `/usr/local/xwiki`, `/opt/solr/server/solr/xwiki`. I don't think it's a big deal to expose `/usr/local/tomcat/conf`.

WDYT?

**Comment by zonradkuse** [comment]
> Inside the image the /usr/local/tomcatconf directory would be a symlink to /usr/local/tomcat/conf.

Yes. I got that. Did you actually try this? I think that it might not work since it is just a symlink.

**Comment by zonradkuse** [comment]
Whoops, your edit didn't show up...

I am going to close this now. I would be glad if you documented how this should be done in `README.md`.

**Comment by vmassol** [comment]
> I would be glad if you documented how this should be done in README.md.

Definitely

>  I do accept very much that this might be an overkill and that you don't think that it is necessary. It is unfortunate considering the time I spent, however, that's the risk about PRs without having spoken to you first.

I'm really sorry about that. However what you did is not fully lost, far from it. In the documentation we can include how you modify the server.xml file to support reverse proxy compatibility.

Thanks a lot

**Comment by vmassol** [comment]
ok I've now tested it and it seems you're right, it's not so easy. "docker run" will let you expose host data to the container but not the other way around. So in order to not have to provide the tomcat conf you need to copy it first from the container to the host (using "docker cp").

So the full workflow would be:

- Create a docker container from the xwiki image ("docker create")
  - Example: `docker create --name xwiki xwiki:lts-mysql-tomcat`
- Copy the Tomcat conf from the container to the host to start with some existing conf files
  - Example: `sudo docker cp xwiki:/usr/local/tomcat/conf /tmp/tomcat`
- Modify the Tomcat configuration locally
- Delete the created xwiki container
  - Example: `docker rm xwiki`
- Run the container with the tomcat mount
  - Example= `docker run --net=xwiki-nw --name xwiki -p 8080:8080 -v /tmp/xwiki:/usr/local/xwiki -v /tmp/tomcat:/usr/local/tomcat/conf -e DB_USER=xwiki -e DB_PASSWORD=xwiki -e DB_DATABASE=xwiki -e DB_HOST=mysql-xwiki xwiki:lts-mysql-tomcat`

So the main change from what I was mentioning above is the "docker cp" that is required to get the existing config, and that requires a created container.

**Comment by vmassol** [comment]
Note that, as I've commented on https://jira.xwiki.org/browse/XDOCKER-69?focusedCommentId=103537&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-103537 I believe the jira issue is still valid and we should provide a flag to make it easy to configure XWiki with a reverse proxy setup and/or use SSL.

But the PR would be about doing that and not about exposing server.xml which is already covered by https://github.com/xwiki-contrib/docker-xwiki/blob/master/README.md#configuring-tomcat

@zonradkuse WDYT?

**Comment by zonradkuse** [comment]
Sure. I think I can modify our ansible accordingly. We are not actually using XWiki's own SSL option since we are running multiple services behind one reverse proxy. But a switch that defaults to no-ssl should be fine I think

**Pull request closed without merging**